pageserver/src/tenant/timeline.rs

```diff
@@ -3271,9 +3271,21 @@ impl Timeline {
                     let mut updates = layers.batch_update();
                     {
                         let l: Arc<dyn PersistentLayer> = remote_layer.clone();
-                        updates.remove_historic(l);
+                        match updates.replace_historic(&l, new_layer) {
+                            Ok(true) => { /* expected */ }
+                            Ok(false) => {
+                                // TODO: the downloaded file should probably be removed, otherwise
+                                // it will be added to the layermap on next load? we should
+                                // probably restart any get_reconstruct_data search as well.
+                                warn!("replacing downloaded layer into layermap failed because layer was not found");
```

Comment:
> If we fail to replace the layer, the next access will retry the on-demand download.

Ending up in the `Ok(false)` would mean that the layer we downloaded was no longer in the layermap. 

I think this retrying could only happen when there's another task already queued up by the semaphore waiting for the first attempt to fail to try downloading it again, but we don't diverge here, so the semaphore is closed and ... Something happens afterwards. I think the search should be restarted but I didn't find any way to signal that.

But this is exactly what I think should be handled on a follow-up (hence TODO).

Reply:
Yeah, not debating any of that. Still, it should be `error!()`, not `warn!()`. The resident size metric will be off in case there's a retry.

Reply:
Changing this to `error!` as wished after rebase, also linking the created todo issue. I don't however think that this should be an error, at least per the quoted rfc:

- [x] not operating normally because we don't see how this could happen during normal operation
- [x] yes, unexpected
- [x] we continue anyway as if the replacement had worked (no change here)

I think errors were meant to be used only with error handling (likely on the TODO issue), which in this case will most likely be the get_reconstruct_data not finding anything, and postgres ending up looking at a zero page.